### PR TITLE
included feature list as input AnnotationManager

### PR DESCRIPTION
**Pull request by lisakoolen**
Thank you for your contribution to the **Tomni** package.
Before submitting/accepting this Pull Request, please make sure:

## Contributor of code

- [ ] Version number is updated
- [ ] `HISTORY.rst` is updated
- [ ] Function is tested
- [ ] `(docstring)` is added or updated
- [ ] All needed package are in [requirements.txt](requirements.txt)

**For new function**

- [ ] Functions are named based on [Python naming conventions](https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html)
- [ ] Function is added to apidocx.rst
- [ ] New feature is added to `README.md` subsection `Features`
- [ ] New function is imported in all `__init__.py` that are needed

**For bugfix**
- [ ] Unit test is added that covers the bug (if possible)
- [ ] You cursed at the person that introduced the bug

## Reviewer

- [ ] Version number is updated
- [ ] `HISTORY.rst` is updated
- [ ] Function is tested
- [ ] `(docstring)` is added or updated
- [ ] All needed package are in [requirements.txt](requirements.txt)

**For new function**

- [ ] Functions are named based on [Python naming conventions](https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html)
- [ ] Function is added to apidocx.rst
- [ ] New feature is added to `README.md` subsection `Features`
- [ ] New function is imported in all `__init__.py` that are needed

**For bugfix**
- [ ] Unit test is added that covers the bug (if possible)

### CODE REVIEW

File: tomni/annotation_manager/annotations/ellipse/main.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def __init__(
         children: List[Annotation] = [],
         parents: List[Annotation] = [],
         radius_y: Union[float, None] = None,
+        features: List[str] = None,
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/ellipse/main.py:24]:
```suggestion
        features: Union[List[str], None] = None,
```

File: tomni/annotation_manager/annotations/ellipse/main.py [hunk 2]
```diff
@@ -37,6 +38,8 @@ def __init__(
             label (str): Class label of annotation.
             children (List[Annotation]): Tracking annotations. Refers to t+1.
             parents (List[Annotation]): Tracking annotations. Refers to t-1.
+            features (List[str]): list of features that the user wants returned.
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/ellipse/main.py:41]:
```suggestion
            features (Union[List[str], None]): list of features that the user wants returned.
```

File: tomni/annotation_manager/annotations/ellipse/main.py [hunk 3]
```diff
@@ -47,10 +50,36 @@ def __init__(
...
+                # check if the input features exist
+                check_feature_possibility = all(
+                    item in all_possible_features for item in features
+                )
+                if check_feature_possibility == False:
+                    raise ValueError(
+                        "One or more features do not exist. Check the spelling of the input features or if that feature is implemented."
+                    )
+
+            self._features = features
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/ellipse/main.py:75]:
```suggestion
        # Determine features to be calculated
        all_features = [
            "area",
            "circularity",
            "perimeter",
            "aspect_ratio",
        ]
        if features is None:
            self._features = all_features
        else:
            missing_features = set(features) - set(all_features)
            if missing_features:
                raise ValueError(f"The following features are not compatible with the Annotation Manager: {', '.join(missing_features)}")
            self._features = features
```

File: tomni/annotation_manager/annotations/polygon/main.py [hunk 4]
```diff
@@ -107,34 +154,47 @@ def points(self, *arg, **kwargs) -> None:
...
-                "convex_hull_area": round(self.convex_hull_area, decimals),
-                "perimeter": round(self.perimeter, decimals),
-                "roundness": round(self.roundness, decimals),
-            }
-            polygon_dict = {**polygon_features, **polygon_dict}
+            if self._has_enough_points:
+                # these statements calculate the features if they are not None. If the feature is None nothing is calculated.
+                # it looks like these arent used but the line: polygon_features[feature] = round(eval(feature), decimals)
+                # uses it when the corresponding name is included in the feature list
+                area = self.area
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/polygon/main.py:186]: if u use getattri, u dont need this afaik

File: tomni/annotation_manager/annotations/polygon/main.py [hunk 5]
```diff
@@ -107,34 +154,47 @@ def points(self, *arg, **kwargs) -> None:
...
+                convex_hull_area = self.convex_hull_area
+                perimeter = self.perimeter
+                circularity = self.circularity
+                roundness = self.roundness
+
+                # when the corresponding name occurs in the feature list, it is added to the dict polygon features
+                polygon_features = {}
+                for feature in self._features:
+                    polygon_features[feature] = round(eval(feature), decimals)
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/polygon/main.py:195]:
```suggestion
                for feature in self._features:
                    polygon_features[feature] = round(getattr(self, feature), decimals)
```

File: tomni/annotation_manager/annotations/ellipse/main.py [hunk 6]
```diff
@@ -153,12 +182,17 @@ def to_dict(self, decimals: int = 2, **kwargs) -> dict:
...
         }
 
+        if self._features:
+            aspect_ratio = self.aspect_ratio
+            area = self.area
+            circularity = self.circularity
+            perimeter = self.perimeter
+
+            for feature in self._features:
+                dict_ellipse[feature] = round(eval(feature), decimals)
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/ellipse/main.py:194]:
```suggestion
            for feature in self._features:
                dict_ellipse[feature] = round(getattr(self, feature), decimals)
```

File: tomni/annotation_manager/annotations/ellipse/main.py [hunk 7]
```diff
@@ -153,12 +182,17 @@ def to_dict(self, decimals: int = 2, **kwargs) -> dict:
...
-            "area": round(self.area, decimals),
-            "circularity": round(self.circularity, decimals),
-            "perimeter": round(self.perimeter, decimals),
         }
 
+        if self._features:
+            aspect_ratio = self.aspect_ratio
+            area = self.area
+            circularity = self.circularity
+            perimeter = self.perimeter
```
Comment (Bram-van-der-Velden) [inline tomni/annotation_manager/annotations/ellipse/main.py:191]: remove

**Comment by Bram-van-der-Velden** [comment]
Unit test fail

**Comment by Bram-van-der-Velden** [comment]
it only init the feature list for "from_dicts"
has to be added to all "from_xxx" functions. e.g. from_contours

**Review by jellevankerk — APPROVED** [review]

**Pull request merged**